proof/invariant-abstract/AARCH64/ArchVCPUDomain_AI.thy

```diff
@@ -0,0 +1,597 @@
+  (wp: cur_vcpu_in_cur_domain_lift)
+
+crunch
+  perform_vspace_invocation, perform_page_table_invocation, perform_page_invocation,
+  perform_asid_control_invocation, perform_asid_pool_invocation, perform_sgi_invocation
+  for cur_vcpu_in_cur_domain[wp]: cur_vcpu_in_cur_domain
+  (wp: crunch_wps  simp: crunch_simps)
+
+lemma arch_perform_invocation_valid_cur_vcpu[wp]:
+  "\<lbrace>cur_vcpu_in_cur_domain and ct_in_cur_domain and cur_vcpu and valid_arch_inv ai
+    and (\<lambda>s. scheduler_action s = resume_cur_thread)\<rbrace>
+   arch_perform_invocation ai
+   \<lbrace>\<lambda>_. cur_vcpu_in_cur_domain\<rbrace>"
+  unfolding arch_perform_invocation_def
+  by (wpsimp simp: valid_arch_inv_def)
+
+\<comment> \<open>FIXME: move (where?)\<close>
+lemma bound_vcpu_bound_tcb_at:
+  "\<lbrakk>sym_refs (state_hyp_refs_of s); valid_objs s; vcpu_tcbs_of s vcpuptr = Some tcbptr\<rbrakk>
+   \<Longrightarrow> arch_tcb_at (\<lambda>itcb. itcb_vcpu itcb = (Some vcpuptr)) tcbptr s"
+  apply (drule_tac x=vcpuptr in sym_refsD[rotated])
+   apply (fastforce simp: state_hyp_refs_of_def pred_tcb_at_def obj_at_def in_omonad)
+  apply (clarsimp simp: in_omonad)
+  apply (erule (1) valid_objsE)
+  apply (clarsimp simp: valid_obj_def valid_vcpu_def typ_at_eq_kheap_obj)
+  apply (fastforce simp: pred_tcb_at_def obj_at_def state_hyp_refs_of_def tcb_vcpu_refs_def
+                  split: option.splits)
+  done
+
+lemma thread_set_domain_cur_vcpu_in_cur_domain[wp]:
+  "\<lbrace>\<lambda>s. cur_vcpu_in_cur_domain s \<and> sym_refs (state_hyp_refs_of s) \<and> valid_objs s
+        \<and> (arm_current_vcpu (arch_state s) = None
+           \<or> arch_tcb_at (\<lambda>itcb. itcb_vcpu itcb \<noteq> map_option fst (arm_current_vcpu (arch_state s))) tptr s
+           \<or> cur_domain s = new_dom)\<rbrace>
+   thread_set_domain tptr new_dom
+   \<lbrace>\<lambda>_. cur_vcpu_in_cur_domain\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def
+  apply (wpsimp wp: set_object_wp)
+  by (auto simp: cur_vcpu_in_cur_domain_defs pred_tcb_at_def obj_at_def etcb_at_def etcbs_of'_def
+          split: option.splits dest: bound_vcpu_bound_tcb_at)
+
+lemma set_domain_cur_vcpu_in_cur_domain[wp]:
+  "\<lbrace>\<lambda>s. cur_vcpu_in_cur_domain s \<and> sym_refs (state_hyp_refs_of s) \<and> valid_objs s
+        \<and> (arm_current_vcpu (arch_state s) = None
+           \<or> arch_tcb_at (\<lambda>itcb. itcb_vcpu itcb \<noteq> map_option fst (arm_current_vcpu (arch_state s))) tptr s
+           \<or> cur_domain s = new_dom)\<rbrace>
+   set_domain tptr new_dom
+   \<lbrace>\<lambda>_. cur_vcpu_in_cur_domain\<rbrace>"
+  unfolding set_domain_def
+  by (wpsimp wp: thread_set_domain_cur_vcpu_in_cur_domain hoare_vcg_disj_lift | wps)+
+
+crunch fpu_release
+  for arch_tcb_vcpu_at[wp]: "\<lambda>s. (arch_tcb_at (\<lambda>itcb. P (itcb_vcpu itcb)) t s)"
+  (simp: crunch_simps wp: crunch_wps)
+
+lemma vcpu_flush_if_current_arm_current_vcpu_None[wp]:
```

Comment:
it looks like this isn't a good name, the postcondition is a simplified implies that says "if it's not none, then it's not the current vcpu"